### PR TITLE
TarReaderStage only pull when needed, #2714

### LINKED ISSUE #2714: TarReaderStage - Cannot pull port (flowIn) twice
Alpakka 3.0.2

```
Error in stage [akka.stream.alpakka.file.impl.archive.TarReaderStage$$anon$1]: Cannot pull port (flowIn(257368048)) twice
java.lang.IllegalArgumentException: Cannot pull port (flowIn(257368048)) twice
        at akka.stream.stage.GraphStageLogic.pull$$original(GraphStage.scala:504)
        at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:29)
        at akka.stream.alpakka.file.impl.archive.TarReaderStage$$anon$1$ExpectDownstreamPull.onPull(TarReaderStage.scala:140)
        at akka.stream.alpakka.file.impl.archive.TarReaderStage$$anon$1$ExpectDownstreamPull.onPull$(TarReaderStage.scala:139)
        at akka.stream.alpakka.file.impl.archive.TarReaderStage$$anon$1$CollectHeader.onPull(TarReaderStage.scala:149)
        at akka.stream.impl.fusing.GraphInterpreter.processPull$$original(GraphInterpreter.scala:551)
        at akka.stream.impl.fusing.GraphInterpreter.processPull(GraphInterpreter.scala:54)
```

### PR DESCRIPTION
**Pull request by patriknw**
* to avoid pull twice bugs
* maybe too defensive in some of the cases but since
  the handler switching and pulling from several places
  is rather difficult to follow it's better to be safe than sorry

Fixes #2714

### CODE REVIEW

**Review by ennru — APPROVED** [review]
Agreed, the state switching makes it hard to follow.
LGTM.

**Comment by fernandoracca** [comment]
We have positive reports that this addresses an issue we've encountered of trying to pull data twice.  👍🏼

**Comment by patriknw** [comment]
@fernandoracca Thanks for confirming. We have released 3.0.3, including this.
